x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Rate.java

```diff
@@ -0,0 +1,162 @@
+        Source source = Source.readFrom(in);
+        Expression field = in.readExpression();
+        Expression timestamp = in.readExpression();
+        Expression unit = in.readBoolean() ? in.readExpression() : null;
+        return new Rate(source, field, timestamp, unit);
+    }
+
+    public static void writeRate(PlanStreamOutput out, Rate rate) throws IOException {
+        rate.source().writeTo(out);
+        out.writeExpression(rate.field());
+        out.writeExpression(rate.timestamp);
+        if (rate.unit != null) {
+            out.writeBoolean(true);
+            out.writeExpression(rate.unit);
+        } else {
+            out.writeBoolean(false);
+        }
+    }
+
+    @Override
+    protected NodeInfo<Rate> info() {
+        return NodeInfo.create(this, Rate::new, field(), timestamp, unit);
+    }
+
+    @Override
+    public Rate replaceChildren(List<Expression> newChildren) {
+        final Expression unit = newChildren.size() >= 3 ? newChildren.get(2) : null;
+        return new Rate(source(), newChildren.get(0), newChildren.get(1), unit);
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.DOUBLE;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        TypeResolution resolution = isType(
+            field(),
+            dt -> dt == DataType.COUNTER_LONG || dt == DataType.COUNTER_INTEGER || dt == DataType.COUNTER_DOUBLE,
+            sourceText(),
+            FIRST,
+            "counter_long",
+            "counter_integer",
+            "counter_double"
+        );
+        if (unit != null) {
+            resolution = resolution.and(
+                isType(unit, dt -> dt.isInteger() || EsqlDataTypes.isTemporalAmount(dt), sourceText(), SECOND, "time_duration")
+            );
+        }
+        return resolution;
+    }
+
+    long unitInMillis() {
+        if (unit == null) {
+            return DEFAULT_UNIT.millis();
+        }
+        if (unit.foldable() == false) {
+            throw new IllegalArgumentException("function [" + sourceText() + "] has invalid unit [" + unit.sourceText() + "]");
+        }
+        final Object foldValue;
+        try {
+            foldValue = unit.fold();
+        } catch (Exception e) {
+            throw new IllegalArgumentException("function [" + sourceText() + "] has invalid unit [" + unit.sourceText() + "]");
+        }
+        if (foldValue instanceof Duration duration) {
+            return duration.toMillis();
+        } else {
+            throw new IllegalArgumentException("function [" + sourceText() + "] has invalid unit [" + unit.sourceText() + "]");
+        }
+    }
+
+    @Override
+    public AggregatorFunctionSupplier supplier(List<Integer> inputChannels) {
+        if (inputChannels.size() != 2 && inputChannels.size() != 3) {
+            throw new IllegalArgumentException("rate() requires two or three channels; got " + inputChannels);
```

Comment:
Nit: is it common to reference channels in errors? This seems somewhat cryptic to the user, can it be reworded to better reflect what's wrong with the input?